services/telemetry/telemetry.go

```diff
@@ -67,54 +75,99 @@ const (
 	TRACK_PLUGINS  = "plugins"
 )
 
-// declaring this as var to allow overriding in tests
-var SENTRY_DSN = "placeholder_sentry_dsn"
+type ServerIface interface {
+	Config() *model.Config
+	IsLeader() bool
+	GetPluginsEnvironment() *plugin.Environment
+	License() *model.License
+	GetRoleByName(string) (*model.Role, *model.AppError)
+	GetSchemes(string, int, int) ([]*model.Scheme, *model.AppError)
+}
+
+type TelemetryService struct {
+	srv                        ServerIface
+	dbStore                    store.Store
+	searchEngine               *searchengine.Broker
+	log                        *mlog.Logger
+	segmentClient              analytics.Client
+	rudderClient               rudder.Client
+	TelemetryID                string
+	timestampLastTelemetrySent time.Time
+}
 
-func (s *Server) SendDailyDiagnostics() {
-	s.sendDailyDiagnostics(false)
+func New(srv ServerIface, dbStore store.Store, searchEngine *searchengine.Broker, log *mlog.Logger) *TelemetryService {
+	service := &TelemetryService{
+		srv:          srv,
+		dbStore:      dbStore,
+		searchEngine: searchEngine,
+		log:          log,
+	}
+	service.ensureTelemetryID()
+	service.initSegmentIO("")
+	service.initRudder(RUDDER_DATAPLANE_URL)
+	return service
 }
 
-func (s *Server) sendDailyDiagnostics(override bool) {
-	if *s.Config().LogSettings.EnableDiagnostics && s.IsLeader() && (!strings.Contains(SEGMENT_KEY, "placeholder") || override) {
-		s.initDiagnostics("")
-		s.trackActivity()
-		s.trackConfig()
-		s.trackLicense()
-		s.trackPlugins()
-		s.trackServer()
-		s.trackPermissions()
-		s.trackElasticsearch()
-		s.trackGroups()
-		s.trackChannelModeration()
-	}
-
-	if *s.Config().LogSettings.EnableDiagnostics && s.IsLeader() && ((!strings.Contains(RUDDER_KEY, "placeholder") && !strings.Contains(RUDDER_DATAPLANE_URL, "placeholder")) || override) {
-		s.initRudder(RUDDER_DATAPLANE_URL)
-		s.trackActivity()
-		s.trackConfig()
-		s.trackLicense()
-		s.trackPlugins()
-		s.trackServer()
-		s.trackPermissions()
-		s.trackElasticsearch()
-		s.trackGroups()
-		s.trackChannelModeration()
+func (ds *TelemetryService) ensureTelemetryID() {
```

Comment:
Shouldn't we return an error here? The method logic seemed to be have different paths on different circumstances. At least might want to log at debug when calling this method.

Reply:
This was the previous existing behavior but I think is a good thing to at least, log it.

Reply:
Done